From 43baf90f96fb9df607595b63858fd9e85fbb2604 Mon Sep 17 00:00:00 2001
From: Wenju He <wenju.he@intel.com>
Date: Sat, 5 Aug 2023 00:33:36 +0800
Subject: [PATCH 148/172] Fix delete of functions that becomes unused (#2109)
 (#2114)

After the first loop of deleting instructions in ValuesToDelete, deleted
instructions in ValuesToDelete are in an unstable state. Then in the
second loop of deleting, dyn_cast to GlobalValue could return true for
an instruction and double eraseFromParent causes crash.

Global values in ValuesToDelete are functions. Unused functions are
deleted by eraseUselessFunctions anyway.

(cherry picked from commit aea1ac74ed4301df12adbffff535f72f37d0bb13)
---
 lib/SPIRV/OCLToSPIRV.cpp | 9 ++-------
 lib/SPIRV/OCLToSPIRV.h   | 2 +-
 2 files changed, 3 insertions(+), 8 deletions(-)

diff --git a/lib/SPIRV/OCLToSPIRV.cpp b/lib/SPIRV/OCLToSPIRV.cpp
index 0772386a..9084c00f 100644
--- a/lib/SPIRV/OCLToSPIRV.cpp
+++ b/lib/SPIRV/OCLToSPIRV.cpp
@@ -159,12 +159,8 @@ bool OCLToSPIRVBase::runOCLToSPIRV(Module &Module) {
 
   visit(*M);
 
-  for (auto &I : ValuesToDelete)
-    if (auto Inst = dyn_cast<Instruction>(I))
-      Inst->eraseFromParent();
-  for (auto &I : ValuesToDelete)
-    if (auto GV = dyn_cast<GlobalValue>(I))
-      GV->eraseFromParent();
+  for (Instruction *I : ValuesToDelete)
+    I->eraseFromParent();
 
   eraseUselessFunctions(M); // remove unused functions declarations
   LLVM_DEBUG(dbgs() << "After OCLToSPIRV:\n" << *M);
@@ -1135,7 +1131,6 @@ bool OCLToSPIRVBase::eraseUselessConvert(CallInst *CI, StringRef MangledName,
                     << *CI->getArgOperand(0) << '\n');
     CI->replaceAllUsesWith(CI->getArgOperand(0));
     ValuesToDelete.insert(CI);
-    ValuesToDelete.insert(CI->getCalledFunction());
     return true;
   }
   return false;
diff --git a/lib/SPIRV/OCLToSPIRV.h b/lib/SPIRV/OCLToSPIRV.h
index 8707ad88..79f9d859 100644
--- a/lib/SPIRV/OCLToSPIRV.h
+++ b/lib/SPIRV/OCLToSPIRV.h
@@ -267,7 +267,7 @@ private:
   Module *M;
   LLVMContext *Ctx;
   unsigned CLVer; /// OpenCL version as major*10+minor
-  std::set<Value *> ValuesToDelete;
+  std::set<Instruction *> ValuesToDelete;
   OCLTypeToSPIRVBase *OCLTypeToSPIRVPtr;
 
   ConstantInt *addInt32(int I) { return getInt32(M, I); }
-- 
2.20.1

